packages/ethereum/contracts/src/Crypto.sol

```diff
@@ -0,0 +1,747 @@
+error InvalidCurvePoint();
+error GeneralError();
+error MessageTooLong();
+
+abstract contract HoprCrypto {
+  // secp256k1: y^2 = x^3 + ax + b (mod F_p)
+
+  uint256 constant SECP256K1_B = 0x0000000000000000000000000000000000000000000000000000000000000007;
+  // Field order created by secp256k1 curve
+  uint256 constant SECP256K1_FIELD_ORDER = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141;
+
+  // 
+  uint256 constant SECP256K1_BASE_FIELD_ORDER = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F;
+  // x-component of base point of secp256k1 curve
+  uint256 constant SECP256K1_BASE_POINT_X_COMPONENT = 0x79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798;
+
+  // encoded sign of y-component of base point of secp256k1 curve
+  uint8 constant SECP256K1_BASE_POINT_Y_COMPONENT_SIGN = 27;
+
+  /**
+   * Holds a compact ECDSA signature, following ERC-2098
+   */
+  struct CompactSignature {
+    bytes32 r;
+    bytes32 vs;
+  }
+
+  struct CurvePoint {
+    uint256 x;
+    uint256 y;
+  }
+
+  /**
+   * Checks whether given value is an element of the secp256k1 field
+   */
+  modifier isFieldElement(uint256 el) {
+    if (0 == el || el >= SECP256K1_FIELD_ORDER) {
+      revert InvalidFieldElement();
+    }
+    _;
+  }
+
+  /**
+   * Checks whether given coordinates fulfill the secp256k1 curve equation
+   */
+  modifier isCurvePoint(CurvePoint memory p)  {
+    if ( 
+      mulmod(p.y, p.y, SECP256K1_BASE_FIELD_ORDER) != 
+      (mulmod(mulmod(p.x, p.x, SECP256K1_BASE_FIELD_ORDER), p.x, SECP256K1_BASE_FIELD_ORDER) + 7) % SECP256K1_BASE_FIELD_ORDER
+    ) {
+      revert InvalidCurvePoint();
+    }
+    _;
+  }
+
+  /**
+   * Ticket redemption uses an asymmetric challenge-response mechanism whose verification
+   * requires scalar multiplication of a secp256k1 curve point.
+   *
+   * Due to the lack of a cheap secp256k1 ECMUL precompile, the construction misuses
+   * the ECRECOVER precompile to compute the scalar multiplication over secp256k1.
+   * Although this returns an Ethereum address, the result is usable to validate the response
+   * against the stated challenge.
+   *
+   * For more information see
+   * https://ethresear.ch/t/you-can-kinda-abuse-ecrecover-to-do-ecmul-in-secp256k1-today/2384
+   *
+   * @param scalar to multiply with secp256k1 base point
+   */
+  function scalarTimesBasepoint(uint256 scalar) public pure returns (address) {
+    return
+      ecrecover(
+        0,
+        SECP256K1_BASE_POINT_Y_COMPONENT_SIGN,
+        bytes32(SECP256K1_BASE_POINT_X_COMPONENT),
+        bytes32(mulmod(scalar, SECP256K1_BASE_POINT_X_COMPONENT, SECP256K1_FIELD_ORDER))
+      );
+  }
+
+  /**
+   * Computes multiplicative inverse of secp256k1 field element.
+   *
+   * Uses little Fermat because expmod precompile is more gas-efficient
+   * than extended euclidian algorithm in Solidity.
+   */
+  function invMod(uint256 el) public view returns (uint256 o) {
+    if (el == 0) {
```

Comment:
Are you assuming that `el` is already reduced ?

Reply:
removed the method